rpcs3/Emu/Cell/Modules/cellVoice.cpp

```diff
@@ -1,27 +1,108 @@
-#include "stdafx.h"
+﻿#include "stdafx.h"
 #include "Emu/System.h"
 #include "Emu/Cell/PPUModule.h"
+#include "Emu/IdManager.h"
+#include "Emu/Cell/lv2/sys_event.h"
 
 #include "cellVoice.h"
 
 LOG_CHANNEL(cellVoice);
 
+template <>
+void fmt_class_string<CellVoiceError>::format(std::string& out, u64 arg)
+{
+	format_enum(out, arg, [](CellVoiceError value)
+	{
+		switch (value)
+		{
+			STR_CASE(CELL_VOICE_ERROR_ADDRESS_INVALID);
+			STR_CASE(CELL_VOICE_ERROR_ARGUMENT_INVALID);
+			STR_CASE(CELL_VOICE_ERROR_CONTAINER_INVALID);
+			STR_CASE(CELL_VOICE_ERROR_DEVICE_NOT_PRESENT);
+			STR_CASE(CELL_VOICE_ERROR_EVENT_DISPATCH);
+			STR_CASE(CELL_VOICE_ERROR_EVENT_QUEUE);
+			STR_CASE(CELL_VOICE_ERROR_GENERAL);
+			STR_CASE(CELL_VOICE_ERROR_LIBVOICE_INITIALIZED);
+			STR_CASE(CELL_VOICE_ERROR_LIBVOICE_NOT_INIT);
+			STR_CASE(CELL_VOICE_ERROR_RESOURCE_INSUFFICIENT);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_ATTACHED);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_DETACHED);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_HANDLE);
+			STR_CASE(CELL_VOICE_ERROR_SERVICE_NOT_FOUND);
+			STR_CASE(CELL_VOICE_ERROR_SHAREDMEMORY);
+			STR_CASE(CELL_VOICE_ERROR_TOPOLOGY);
+		}
+		return unknown;
+	});
+}
 
-s32 cellVoiceConnectIPortToOPort()
+struct voice_t
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
+	u64 eventQueueKey = 0;
+	u64 source = 0;
+	bool voiceReady = false;
+};
+
+error_code cellVoiceConnectIPortToOPort(u32 iPort, u32 oPort)
+{
+	cellVoice.todo("cellVoiceConnectIPortToOPort(iPort=0x%x, oPort=0x%x)", iPort, oPort);
+
+	const auto voice = fxm::get<voice_t>();
+	if (!voice)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
 	return CELL_OK;
 }
 
-s32 cellVoiceCreateNotifyEventQueue()
+error_code cellVoiceCreateNotifyEventQueue(vm::ptr<u32> id, vm::ptr<u64> key)
 {
-	UNIMPLEMENTED_FUNC(cellVoice);
-	return CELL_OK;
+	cellVoice.todo("cellVoiceCreateNotifyEventQueue(id=*0x%x, key=*0x%x)", id, key);
+
+	const auto voice = fxm::get<voice_t>();
+	if (!voice)
+	{
+		return CELL_VOICE_ERROR_LIBVOICE_NOT_INIT;
+	}
+
+	// axed from cellAudio
+	vm::var<sys_event_queue_attribute_t> attr;
+	attr->protocol = SYS_SYNC_FIFO;
+	attr->type = SYS_PPU_QUEUE;
+	attr->name_u64 = 0;
+
+	for (u64 i = 0; i < 100; i++)
```

Comment:
According to the firmware, unlike cellAudioCreateNotifyEventQueue, it only tries 10 times until it runs out of keys.